gax-grpc/src/main/java/com/google/api/gax/grpc/SafeShutdownManagedChannel.java

```diff
@@ -0,0 +1,165 @@
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A {@link ManagedChannel} that will complete all calls started on the underlying channel before
+ * shutting down.
+ *
+ * <p>This class is not thread-safe. Caller must synchronize in order to ensure no new calls if safe
+ * shutdown has started.
+ *
+ * <p>Package-private for internal use.
+ */
+class SafeShutdownManagedChannel extends ManagedChannel {
+  private ManagedChannel delegate;
+  private AtomicInteger outstandingCalls = new AtomicInteger(0);
+  private volatile boolean isShutdownSafely = false;
+
+  SafeShutdownManagedChannel(ManagedChannel managedChannel) {
+    this.delegate = managedChannel;
+  }
+
+  /**
+   * Safely shutdown channel by checking that there are no more outstanding calls. If there are
+   * outstanding calls, the last call will invoke this method again when it complete
+   *
+   * <p>Caller should take care to synchronize with newCall so no new calls are started after
+   * shutdownSafely is called
+   */
+  void shutdownSafely() {
+    isShutdownSafely = true;
+    if (outstandingCalls.get() == 0) {
+      delegate.shutdown();
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdown() {
+    delegate.shutdown();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isShutdown() {
+    return delegate.isShutdown();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdownNow() {
+    delegate.shutdownNow();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isTerminated() {
+    return delegate.isTerminated();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return delegate.awaitTermination(timeout, unit);
+  }
+
+  /** Listener that's responsible for decrementing outstandingCalls when the call closes */
+  private class DecrementOutstandingCalls<RespT> extends SimpleForwardingClientCallListener<RespT> {
+    DecrementOutstandingCalls(Listener<RespT> delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public void onClose(Status status, Metadata trailers) {
+      // decrement in finally block in case onClose throws an exception
+      try {
+        super.onClose(status, trailers);
+      } finally {
+        if (outstandingCalls.decrementAndGet() == 0 && isShutdownSafely) {
+          shutdownSafely();
+        }
+      }
+    }
+  }
+
+  /**
+   * Wrap client call to decrement outstandingCalls and shutdown channel if necessary when it
+   * completes
+   *
+   * @param call call to be wrapped
+   * @return the wrapped call
+   */
+  private <ReqT, RespT> ClientCall<ReqT, RespT> clientCallWrapper(ClientCall<ReqT, RespT> call) {
+    return new SimpleForwardingClientCall<ReqT, RespT>(call) {
```

Comment:
I think it would read better if SimpleForwardingClientCall was an inner class and the listener was an anonymous class. Also would be good to keep the business logic of inc/decrementing the ref count on the same level (currently increment happens in the parent and the decrement happens in the inner class

```java
private void onClientCallClose() {
   if (outstandingCalls.decrementAndGet() == 0 && isShutdownSafely) {
          shutdownSafely();
     }
}

public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
      MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
   Preconditions.check...
   ClientCall innerCall = delegate.newCall(methodDescriptor, callOptions);
   outstandingCalls.incrementAndGet()
   return new ClientCallProxy(innerCall);
}

class ClientCallProxy extends SimpleForwardingClientCall {
   start() {
      return new SimpleForwardingClientCallListener() {
             ....
            onClose() {
                onClientCallClose()
            }
      }
   }
}
```